process_scheduling_algorithms/preemptive_priority_scheduling.c

```diff
@@ -0,0 +1,281 @@
+ */
+float calculate_avg_ct(Process *processes, int n) {
+    float sum = 0;
+    for (int i = 0; i < n; i++) {
+        sum += processes[i].finishTime;
+    }
+    return sum / n;
+}
+
+/**
+ * @brief Calculates the average turnaround time of all the processes
+ * @param processes array of processes
+ * @param n number of processes
+ * @returns float average turnaround time
+ */
+float calculate_avg_tat(Process *processes, int n) {
+    float sum = 0;
+    for (int i = 0; i < n; i++) {
+        sum += processes[i].turnaroundTime;
+    }
+    return sum / n;
+}
+
+/**
+ * @brief Calculates the average waiting time of all the processes
+ * @param processes array of processes
+ * @param n number of processes
+ * @returns float average waiting time
+ */
+float calculate_avg_wt(Process *processes, int n) {
+    float sum = 0;
+    for (int i = 0; i < n; i++) {
+        sum += processes[i].waitingTime;
+    }
+    return sum / n;
+}
+
+/**
+ * @brief Performs preemptive priority scheduling on the processes
+ * @param processes array of processes
+ * @param n number of processes
+ * @returns void
+ */
+void preemptivePriorityScheduling(Process *processes, int n) {
+    qsort(processes, n, sizeof(Process), compare);
+
+    int currentTime = 0;
+    int completedProcesses = 0;
+
+    while (completedProcesses < n) {
+        int highestPriorityIndex = -1;
+        int highestPriority = 9999;
+
+        for (int i = 0; i < n; i++) {
+            if (processes[i].arrivalTime <= currentTime && processes[i].burstTime > 0 && processes[i].priority < highestPriority) {
+                highestPriorityIndex = i;
+                highestPriority = processes[i].priority;
+            }
+        }
+
+        if (highestPriorityIndex != -1) {
+            processes[highestPriorityIndex].burstTime--;
+            currentTime++;
+
+            if (processes[highestPriorityIndex].startTime == -1) {
+                processes[highestPriorityIndex].startTime = currentTime - 1;
+            }
+
+            if (processes[highestPriorityIndex].burstTime == 0) {
+                completedProcesses++;
+                processes[highestPriorityIndex].finishTime = currentTime;
+            }
+        } else {
+            currentTime++;
+        }
+    }
+
+    calculate_ct(processes, n);
+    calculate_tat(processes, n);
+    calculate_wt(processes, n);
+}
+
+/**
+ * @brief Prints a table of the processes
+ * @param processes array of processes
+ * @param n number of processes
+ * @returns void
+ */
+void printProcessTable(Process *processes, int n) {
+    printf("\nProcess\tArrival Time\tBurst Time\tPriority\tWaiting Time\tTurnaround Time\n");
+    for (int i = 0; i < n; i++) {
+        printf("%s\t%12d\t%10d\t%8d\t%12d\t%15d\n", processes[i].name, processes[i].arrivalTime, processes[i].burstTime, processes[i].priority, processes[i].waitingTime, processes[i].turnaroundTime);
+    }
+}
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+void test() {
```

Comment:
Your `test()` is unused, also it doesn't really test anything since the results are not compared against expectations